### PR TITLE
plan: implement ExplainInfo() interface for join operators

### PR DESCRIPTION
**Pull request by zz-jason**
to https://github.com/pingcap/tidb/issues/3805

### CODE REVIEW

File: plan/explain.go [hunk 1]
```diff
@@ -127,3 +127,254 @@ func (p *PhysicalAggregation) ExplainInfo() string {
 	}
 	return buffer.String()
 }
+
+// ExplainInfo implements PhysicalPlan interface.
+func (p *PhysicalApply) ExplainInfo() string {
```
Comment (hanfei1991) [inline plan/explain.go]: Add test for all `ExplainInfo`

File: plan/explain.go [hunk 2]
```diff
@@ -127,3 +127,254 @@ func (p *PhysicalAggregation) ExplainInfo() string {
...
+	buffer.WriteString(fmt.Sprintf(", right:%s", p.Children()[p.rightChOffset].ID()))
+	return buffer.String()
+}
+
+// ExplainInfo implements PhysicalPlan interface.
+func (p *PhysicalIndexJoin) ExplainInfo() string {
+	buffer := bytes.NewBufferString(fmt.Sprintf("outer:%s", p.Children()[p.outerIndex].ID()))
+	if len(p.OuterJoinKeys) > 0 {
+		buffer.WriteString(", outer key:")
+		for i, col := range p.OuterJoinKeys {
```
Comment (lamxTyler) [inline plan/explain.go]: Could this for-loop be abstracted as a function? It appears so many times.

**Comment by hanfei1991** [comment]
LGTM

**Comment by winoros** [comment]
sqllogic test failed.

File: expression/explain.go [hunk 3]
```diff
@@ -60,3 +60,27 @@ func ExplainAggFunc(agg AggregationFunction) string {
 	buffer.WriteString(")")
 	return buffer.String()
 }
+
+// ExplainExpressionList generates explain information for a list of expressions
```
Comment (lamxTyler) [inline expression/explain.go]: Add `.`.

File: expression/explain.go [hunk 4]
```diff
@@ -60,3 +60,27 @@ func ExplainAggFunc(agg AggregationFunction) string {
...
+	for i, expr := range exprs {
+		buffer.WriteString(expr.ExplainInfo())
+		if i+1 < len(exprs) {
+			buffer.WriteString(", ")
+		}
+	}
+	return buffer.Bytes()
+}
+
+// ExplainColumnList generates explain information for a list of columns
```
Comment (lamxTyler) [inline expression/explain.go]: Add `.`.

**Pull request merged**